### PR TITLE
[Demuxer][InputStream] Add DEMUX_SPECIALID_STREAMCHANGE_AV

### PR DESCRIPTION
**Pull request by CastagnaIT**
## Description

This add a new demux specialid DEMUX_SPECIALID_STREAMCHANGE_AV,
this case works in same way of DEMUX_SPECIALID_STREAMCHANGE but avoid to reinitialize the subtitles streams so the current streams/settings are kept as is until the end of playback

## Motivation and context

Currently DEMUX_SPECIALID_STREAMCHANGE is used to reinitialize all streams in kodi player while playing,
at today InputStream Adaptive has been introduced an adaptive stream switching based on network bandwidth,
then the DEMUX_SPECIALID_STREAMCHANGE callback can happen very often during the playback,
and this call cause problems with subtitle management.

The problem is that every time that this callback is sent from ISA during the playback cause following problems, more or less in this order:
1) ISA want to switch video stream, send DEMUX_SPECIALID_STREAMCHANGE to kodi
2) Kodi reset player streams (subtitles included)
3) Request to update streams from ISA
4) ISA download again the same subtitles data
5) Kodi initialize again a subtitle parser with a new Libass instance and re-parse whole subtitle data, you can see in the log `Created subtitles overlay codec: ***` after `AddOnLog: inputstream.adaptive: DEMUX_SPECIALID_STREAMCHANGE`
6) I think due to the subtitle comparer (PredicateSubtitlePriority) is again evaluated which subtitle track to display on the screen, that could be also disabled, causing a mess, because this was already done on video starts, and the user can have selected a subs track of his preference (it is not always possible to have/show this behavior, but anyway you can see in the log that the subtitle track is reinitialized)

Some other details on this comment: https://github.com/xbmc/inputstream.adaptive/issues/836#issuecomment-977739259

Try fixing #20404

@glennguy for reference

## How has this been tested?

I have modified ISAdaptive to use DEMUX_SPECIALID_STREAMCHANGE_AV (two lines in main.cpp),
and included the PR's (will be merged soon):
https://github.com/xbmc/inputstream.adaptive/pull/841
https://github.com/xbmc/inputstream.adaptive/pull/820
then to make ISA do the DEMUX_SPECIALID_STREAMCHANGE_AV callback's, i have played a Netfx video trailer, and started on a separated browser at same time 2 youtube videos to slow down the network and then ISA buffering.
when the callback happens, you will find it in the kodi log the name `DEMUX_SPECIALID_STREAMCHANGE_AV` printed

How to reproduce the problem with a Dash stream and regular Kodi20 build/ISA build:
- Play this Dash STRM: [_Test_WebVTT_segmented_HLS.strm.txt](https://github.com/xbmc/xbmc/files/7607862/_Test_WebVTT_segmented_HLS.strm.txt)
- Load an external subtitle file and enable subtitles via GUI
- Slow down your network to make ISA changing the video stream quality
- When ISA will change quality `AddOnLog: inputstream.adaptive: DEMUX_SPECIALID_STREAMCHANGE` will be printed on the Kodi log (keep track in realtime) after that you can see also `Created subtitles overlay codec: ***` that means that has been resetted the subtitle parser instance and requested again subtitle data and re-parsed the subtitle data
- Now (always while in playback) open GUI subtitles settings, you'll see that the external subtitles track is no longer selected and it has re-selected Kodi default subtitle language track
None of this should have happened while changing A/V quality stream

## What is the effect on users?

They can watch videos without having subtitle resetted or disabled every time the video stream is switched by ISAdaptive

## Types of change

- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by CastagnaIT** [comment]
I am not sure who could give a opinion on this maybe @djp952 has helped in the past with ISA or @FernetMenta if it is still around

**Comment by djp952** [comment]
I don't think I'd be a very good reviewer for you here, @CastagnaIT, almost everything I've tried to "fix" in VideoPlayer has fallen pretty flat. I think I managed to fix something properly a grand total of once :) @FernetMenta is definitely a good choice IMO, but (also IMO, not trying to speak for him) he seems to feel that a broader refactor of VideoPlayer is in order to provide a more modern implementation that would avoid these problems rather than pick and choose one-off tweaks.

From what I see in the PR, it seems reasonable, and seems you have a way to both reproduce the problem and proof that the PR solves that particular problem.

I have some Blu-Ray based content that has odd problems with subtitles (the original Swedish "Girl with the Dragon Tattoo" films), I would be happy to see if maybe this PR would solve those problems?

From a code review perspective, I question if the new **DMX_SPECIALID_STREAMCHANGE_AV** flag is necessary or if your updated logic for dealing with the subtitle stream could just be applicable by default?  If the existing **DMX_SPECIALID_STREAMCHANGE** flag would work as you think it should if it ignored **STREAM_SUBTITLE**, or maybe only cared about **STREAM_AUDIO | STREAM_VIDEO** (and perhaps **STREAM_DATA**?), would that solve the underlying problem effectively?  The remaining flags **TELETEXT** and **RADIO_RDS** might also be better served by being set once and then ignored, yes?

I apologize if this feedback is without any value-add.

**Comment by FernetMenta** [comment]
> Not knowing how this streams ecosystem works, i'm trying to propose a possible solution

I think this is the wrong approach. In order to fix something you should get familiar with how the system is supposed to work.

> Currently DEMUX_SPECIALID_STREAMCHANGE is used to reinitialize all streams in kodi player while playing

This assumption is wrong. DEMUX_SPECIALID_STREAMCHANGE signals a stream change and VP is supposed to adapt to the change with minimal impact. DEMUX_SPECIALID_STREAMCHANGE is fired often on PVR streams in particular on commercial breaks. In general the video stream doesn't change in this case and VP won't touch it. The same should apply for subtitles. If this is not the case, something goes wrong and I would rather fix this than implement some additional code path.

**Comment by CastagnaIT** [comment]
> I think this is the wrong approach. In order to fix something you should get familiar with how the system is supposed to work.

seems to be short of developers in this area, so i came forward to propose/find a solution, thats why i have marked as RFC,
if this is a problem i can close the PR, hoping that someone will take care of this, but also in InputStream Adaptive project we have only one active dev and no has much time to take care of everything

The InputStream Adaptive case, is a bit different than PVR,
i am not able to test PVR but i think that commercial breaks can change subtitles or have no subtitles, then the current behaviour can be right for that case, the current code assumes that subtitles streams may differ on each call, and are every time re-evaluated and re-opened/closed.

But ISA call DEMUX_SPECIALID_STREAMCHANGE to signal a change in video/audio bitrate (maybe also video resolution not sure implemented ATM) and in this case the subtitle streams are kept the same and no need to be re-evaluated and re-opened/closed.
This is the reason behind whitch i add a new DEMUX_SPECIALID_STREAMCHANGE to exclude the continuous evaluation with
`PredicateSubtitlePriority` selector, then `OpenStream`/`CloseStream` cause to open/close the subtitle stream and then reinitialization of Kodi parser/Libass and after that `SetSubtitleVisibleInternal` call in some cases can disable the subtitles.

But perhaps an alternative approach (to avoid add the DEMUX_SPECIALID_STREAMCHANGE_AV) is to make optional that block of code by using something similar to `CVideoPlayer::CheckStreamChanges` to check whether the full list of subtitle streams properties has changed, then execute that block of code only when something in the subtitle stream list is changed

**Comment by CastagnaIT** [comment]
> From a code review perspective, I question if the new DMX_SPECIALID_STREAMCHANGE_AV flag is necessary or if your updated logic for dealing with the subtitle stream could just be applicable by default?

there is no a default, because depends on the needs, the choice falls on InputStream add-on demuxers who will use the most appropriate.

I have get a confirm from a CoreElec dev that have tested PVR, the `DMX_SPECIALID_STREAMCHANGE` is called when switch channel or switch channel to a recorded one, so the current (original) behaviour is fine for PVR case

but not fit the InputStreamAdaptive use,
because we need to keep: the same subtitle track / same language setting and keep alive the subtitle codec handler even while we have the adaptive stream enabled (automatic switching between audio/video stream qualities while watching a video)

I have reworked the code i think better i ask for second review

File: xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemux.h [hunk 1]
```diff
@@ -47,7 +47,8 @@ enum StreamType
   STREAM_DATA, // data stream
   STREAM_SUBTITLE, // subtitle stream
   STREAM_TELETEXT, // Teletext data stream
-  STREAM_RADIO_RDS // Radio RDS data stream
+  STREAM_RADIO_RDS, // Radio RDS data stream
+  STREAM_ENUM_COUNT
```
Comment (enen92) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemux.h:51]: Is this right here? All the other enum entries refer to a concrete type, what is `STREAM_ENUM_COUNT`?

**Comment by CastagnaIT** [comment]
@FernetMenta some comments above you said this:

> The same should apply for subtitles. If this is not the case, something goes wrong and I would rather fix this than implement some additional code path.

can you explain what you mean by fixing what goes wrong?

because i have tried to find other alternative to fix this problem without any success, because the problem is right here in these lines: https://github.com/xbmc/xbmc/blob/093c88eeee6c6d0bc8cb0c8a5c1ae64641626972/xbmc/cores/VideoPlayer/VideoPlayer.cpp#L931-L953

recap:

these lines of the code assumes that you ALWAYS want to reset the current subtitles codec handler and invalidate the current active subtitle track. Then it close the subtitle, select a new one based on kodi settings, and if find a new track open the new sub stream that may be different from the previous or also not match and so disable it.

Now on ISAdaptive we have adaptive stream feature, this is a new feature never added in kodi,
that switch audio/video stream qualities while watching a video automatically,
i can assume that on the PVR the adaptive stream feature is not implemented, otherwise you will have the exact same problems, with the need to do something similar of this PR.

therefore what you want to achieve is avoid in full execute that block of code

the only other alternative that i thought add a new parameters to DMX_SPECIALID_STREAMCHANGE demuxpacket, in order to allow InputStream to select the types of streamchange

example (from ISAdaptive demux packet sent):

```
DEMUX_PACKET* p = AllocateDemuxPacket(0);
p->iStreamId = DEMUX_SPECIALID_STREAMCHANGE;
p->StreamChangeTypeFlags = STREAMCHANGE_TYPE_VIDEO | STREAMCHANGE_TYPE_AUDIO;
```

or to resetting also subtitles:

```
DEMUX_PACKET* p = AllocateDemuxPacket(0);
p->iStreamId = DEMUX_SPECIALID_STREAMCHANGE;
p->StreamChangeTypeFlags = STREAMCHANGE_TYPE_VIDEO | STREAMCHANGE_TYPE_AUDIO | STREAMCHANGE_TYPE_SUBTITLES;
```

then on VP will only act on the types set, and if this param is not set, will works as before
if you know another better way to make this change i am happy to have some explanation

**Comment by FernetMenta** [comment]
> these lines of the code assumes that you ALWAYS want to reset the current subtitles codec handler and invalidate the current active subtitle track

No, this is not the case, at least not in general. This is what I meant with my comment. Subs should be handled like a/v streams. There is one exception with closed captions because those are muxed into the video stream. I guess this is what this old change tried to fix: https://github.com/xbmc/xbmc/pull/6105
Maybe close subs only in case of cc.

**Comment by CastagnaIT** [comment]
> Maybe close subs only in case of cc.

yes tested by commenting `CloseStream`: https://paste.kodi.tv/saqobusezo.kodi

the subtitle stream (codec handler) is always reopened at each call (by `OpenStream`)
in the log after each DEMUX_SPECIALID_STREAMCHANGE you will find: `CALL TO OpenDefaultStreams`
a print that i added after the commented `CloseStream`
you can see `OpenStream` by `Opening stream: ...` comment and relative codec/parser reinitialization:
`CDVDSubtitlesLibass: Creating ASS library structure`
`Created subtitles overlay codec: WebVTT Subtitle Decoder`

**Comment by FernetMenta** [comment]
not sure what you are trying to say

**Comment by CastagnaIT** [comment]
that in any case `OpenStream` is fired and every time open again the stream by creating/initialize every time from scratch a new codec handler/parser and this must not happens

**Comment by FernetMenta** [comment]
This is a problem of inputstream addon. The design is as follows: VP can call OpenDefaultStreams as many times as it wants to, for what reason ever. This must not do any harm to playing streams if nothing has changed.

**Comment by CastagnaIT** [comment]
> The design is as follows: VP can call OpenDefaultStreams as many times as it wants to, for what reason ever. This must not do any harm to playing streams if nothing has changed.

ok, but at present, this does not happen correctly

at each DEMUX_SPECIALID_STREAMCHANGE, as i have already shown with the log,
the `OpenStream` every time open a new codec hadler/parser
without any checks if the stream used is the same
therefore always initialise a new codec/parser

on slack we have talked about another solution, about to check if the current stream ID exists in the new stream list,
and then only if not exists, execute CloseStream/OpenStream,
the problem in do this is that we do not know if the PVR subtitles streams IDs are always different

**Comment by FernetMenta** [comment]
Consider a very simple API with a function TurnOnLight. If you call this function when the light is already on, you expect nothing to happen. That's the design of OpenStream. Sure, you can store the state information of the subcomponent in the controller (VP in this case) but this just increases complexity at this level. 
Looks like all demuxers but ISA have no issue with OpenStream. You can fix ISA or pollute VP with code only required by ISA because it doesn't follow the original API design.

**Comment by CastagnaIT** [comment]
> Consider a very simple API with a function TurnOnLight. If you call this function when the light is already on, you expect nothing to happen. 

well i'm still saying that this design doesn't work. not as you would like it to be.

sticking to your metaphor, as is now currenlty works in this way:
"If you call this function when the light is already on, ~you expect nothing to happen.~ then turn off and turn on"

Everything is re-initialized audio/video/subtitles, and you can see clearly in the log (DXVA / audio SINK / subtiltle codec are reinitialized)

can be good for audio/video, but if you reinitialize the subtitle codec when is not needed you break it.
currently there is no kind of check to "do nothing" behaviour, always you turn off and on.

continue to state that it is an ISA problem
but in ISA subtitles packets are handled as it should,
and the behaviour change depending on type and format of subtitles, that cannot be handled all in same way:

- streams with file type --> subtitle data is sent in full at first PTS "0", ignored other PTS packet request since they not exists
- streams with segmented type --> subtitle data is sent at each relative PTS

so what happens with kodi core turn off/on?
- File type --> you unload the codec/parser in kodi core, you will lose all subtitle data already parsed and stored in the libass handler, result in broken subtiltes.
- Packet type --> unload the codec/parser in kodi core, for this case this _may_ not matter much, but if libass has several fonts to load in memory at his initilialization, will take a long time and subtitles will not be displayed for a period of time. This will happens at each streamchange demuxpacket then if you have an unstable network you are screwed kodi will broke subtitles every time.

with internet VOD/Live streams the subtitle streams are fixed and not change over the time, then there is no reason to (turn off/on) reset every time the codec handler but still use the same instance

what changes you want to make in ISA i have no idea...

I tried to add a check to compare current ID to the new stream list,
to add the "do nothing" behaviuour, and works,
but you should test for PVR case
https://github.com/CastagnaIT/xbmc/commit/0e907d8dad12465ca2e366bb23051cb8467d78ff

**Comment by CastagnaIT** [comment]
i have updated the commit:
new link: https://github.com/CastagnaIT/xbmc/commit/b6655448d79428b45f8106ad2800105d1e0ae1dd
I think this is a reasonable change of slight impact

i think @ksooo and/or @phunkyfish can give information for DEMUX_SPECIALID_STREAMCHANGE use cases in PVR addons,
and if possible test linked commit here to see regressions with PVR subtitles

However, i expect subtitles on PVR to be usually CC, or other formats but always as PTS packets based, so they should have no problems with this change

**Comment by CastagnaIT** [comment]
ping @phunkyfish just to remember when you have time test commit in comment above, to check if subtitles still works well in PVR/channel switching thanks

**Comment by phunkyfish** [comment]
> ping @phunkyfish just to remember when you have time test commit in comment above, to check if subtitles still works well in PVR/channel switching thanks

So I should only test the commit you referenced and not the content of this PR?

**Comment by CastagnaIT** [comment]
yes only this commit https://github.com/CastagnaIT/xbmc/commit/b6655448d79428b45f8106ad2800105d1e0ae1dd
will replace this PR if works

**Comment by phunkyfish** [comment]
> yes only this commit [CastagnaIT@b665544](https://github.com/CastagnaIT/xbmc/commit/b6655448d79428b45f8106ad2800105d1e0ae1dd) will replace this PR if works

@emveepee @howie-f any chance you can test this commit on Nexus to check it subtitles still work well when switching channels in PVR?

I'm away and have no access to test this and I don't want to hold anything up.

**Comment by howie-f** [comment]
no issues with pvr.hts, but can‘t check iptvsimple/isa

**Comment by phunkyfish** [comment]
Testing with hts should be sufficient, thanks!

**Comment by emveepee** [comment]
> @emveepee @howie-f any chance you can test this commit on Nexus to check it subtitles still work well when switching channels in PVR?

ATSC CCs are embedded in the video so they aren't separate streams so they should continue to be reset with video.

**Comment by CastagnaIT** [comment]
thanks for the tests
i will do a new PR to superseed this one

**Pull request closed without merging**